### PR TITLE
Mono BCL test fixes so XM can run them

### PR DESCRIPTION
**Pull request by chamons**
I've been working on porting all of the BCL tests XI runs to also run on XM.

There is an unrelated chunk of build infrastructure work associated with this in xamarin-macios as well.

Since the XM tests are not defining MOBILE, we are getting a bunch more tests that XI. In a few places however, I've cut them back to match XI.

There were a few files that ran into conflict due to differences between nunitlight and guiunit. I had to add a few explicit usings / simplify assertions to paper over the differences.

### CODE REVIEW

File: mcs/class/System.XML/Test/System.Xml/XmlSecureResolverTests.cs [hunk 1]
```diff
@@ -136,7 +136,7 @@ public void TestAsync ()
 			Assert.AreEqual ("file", resolved.Scheme);
 			var task = sr.GetEntityAsync (resolved, null, typeof (Stream));
 			Assert.That (task.Wait (3000));
-			Assert.IsInstanceOfType (typeof (FileStream), task.Result);
+			Assert.IsTrue (task.Result is Stream);
```
Comment (akoeplinger) [inline mcs/class/System.XML/Test/System.Xml/XmlSecureResolverTests.cs]: why not `is FileStream`? Also can you please include `"Unexpected type: " + task.Result.GetType()` as the third parameter so we know what's wrong when this task fails?

File: mcs/class/System.Xml.Linq/Test/System.Xml.Schema/ExtensionsTest.cs [hunk 2]
```diff
@@ -220,8 +220,8 @@ public void XDocumentAddSchemaInfoValidate ()
 				}
 			}
 
-			AssertHelper.GreaterOrEqual (afterNoOfAttributes, beforeNoOfAttributes, "newAttributes");
-			AssertHelper.GreaterOrEqual (afterNoOfElements, beforeNoOfElements, "newElements");       
+			Assert.IsTrue (afterNoOfAttributes >= beforeNoOfAttributes, "newAttributes");
+			Assert.IsTrue (afterNoOfElements >= beforeNoOfElements, "newElements");
```
Comment (akoeplinger) [inline mcs/class/System.Xml.Linq/Test/System.Xml.Schema/ExtensionsTest.cs]: same here, include afterNoOfElements and beforeNoOfElements in the assert message

File: mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs [hunk 3]
```diff
@@ -53,7 +53,7 @@ public class SourceSwitchTest
 		public void ConstructorNullName ()
 		{
 			SourceSwitch s = new SourceSwitch (null);
-			AssertHelper.IsEmpty (s.DisplayName);
+			Assert.AreEqual (s.DisplayName.Length, 0);
```
Comment (akoeplinger) [inline mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs]: does `Assert.IsEmpty()` work instead? if yes we should get rid of the AssertHelper.

File: mcs/class/System/Test/System.Diagnostics/FileVersionInfoTest.cs [hunk 4]
```diff
@@ -18,6 +18,7 @@
 using System.Text;
 
 using NUnit.Framework;
+using Assert = NUnit.Framework.Assert;
```
Comment (akoeplinger) [inline mcs/class/System/Test/System.Diagnostics/FileVersionInfoTest.cs]: why is this necessary for XM? XI doesn't seem to need it.

**Comment by akoeplinger** [comment]
@chamons left a few comments, nothing major :) please include the detailed description in the commit message too, not just the PR.

**Comment by marek-safar** [comment]
These `using Assert = NUnit.Framework.Assert;` need explanation it should not be needed

File: mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs [hunk 5]
```diff
@@ -444,7 +444,9 @@ public void FromAsync_ArgumentsCheck ()
 			}
 		}
 
+#if !MONOMAC // Randomly fails - https://bugzilla.xamarin.com/show_bug.cgi?id=51255
```
Comment (marek-safar) [inline mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs]: For such cases we tend to use Category, e.g. [Category ("MacNotWorking")]

File: mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs (hunk 3 above)
Comment (spouliot) [inline mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs]: or `Assert.That (s.DisplayName, Is.Empty)` <- but I'm not sure if mono's current nunit accept that syntax

File: mcs/class/System.XML/Test/System.Xml/XmlSecureResolverTests.cs (hunk 1 above)
Comment (chamons) [inline mcs/class/System.XML/Test/System.Xml/XmlSecureResolverTests.cs]: Simple mistake. Will fix.

File: mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs (hunk 3 above)
Comment (akoeplinger) [inline mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs]: @spouliot that's actually what AssertHelper.IsEmpty is doing now since we switched to nunitlite: https://github.com/mono/mono/blob/695282c94ba45c27f909e843ca2b7a86c7bd085a/mcs/class/test-helpers/NunitHelpers.cs#L67-L70

I'm a bit surprised these cause problems for XM since I thought guiunit is nunitlite based too?

File: mcs/class/System/Test/System.Diagnostics/FileVersionInfoTest.cs (hunk 4 above)
Comment (chamons) [inline mcs/class/System/Test/System.Diagnostics/FileVersionInfoTest.cs]: Hmm. I'm not honestly sure now. I was getting compile errors when I added it but now it's building fine. Removing.

File: mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs (hunk 3 above)
Comment (chamons) [inline mcs/class/System/Test/System.Diagnostics/SourceSwitchTest.cs]: So It looks like this changed from when I started my BCL work. This commit is not in 4.8 (where we were tracking in master and I started this):

https://github.com/mono/mono/commit/863dc73d1984175cb52734e1cb755c9d57de53bd

And my resolve kept my changes. These changes now look unnecessary (at least in master, and we can cherry pick that into 4.8 if necessary).

File: mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs (hunk 5 above)
Comment (akoeplinger) [inline mcs/class/corlib/Test/System.Threading.Tasks/TaskFactoryTest.cs]: I'd actually prefer it this way since it makes clear that it doesn't work on XM (instead of Mac in general like `MacNotWorking` implies)

**Comment by akoeplinger** [comment]
@chamons oh that explains a lot, we only migrated from nunit 2.4 to nunitlite in master 😄
